Prevent client skew

deads2k opened the pull request:
Adds an option to prevent skewed oc clients from performing PUT or POST (update or create) actions on the API server.  This can prevent weirdness where fields are dropped by old oc clients.  The default is to avoid restricting any client.  Any client not strictly conforming to `DefaultOpenShiftUserAgent` is allowed.

Also fixed here: update the server and client to always identify as openshift, not as kubernetes.  Our client is not the same as kube and the commits didn't really line up due to patches anyway.

@fabianofranz @jwforres since you asked.
@liggitt hold your nose, I don't think its ever going to smell nice.

Buzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1306590

liggitt commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/helper.go:
```diff
@@ -580,7 +580,7 @@ func DefaultKubernetesUserAgent() string {
 	seg := strings.SplitN(version, "-", 2)
 	version = seg[0]
-	return fmt.Sprintf("%s/%s (%s/%s) kubernetes/%s", path.Base(os.Args[0]), version, gruntime.GOOS, gruntime.GOARCH, commit)
+	return fmt.Sprintf("%s/%s (%s/%s) openshift/%s", path.Base(os.Args[0]), version, gruntime.GOOS, gruntime.GOARCH, commit)
```
origin vs openshift vs aep?

deads2k replied:
> origin vs openshift vs aep?

"openshift" is hardcoded in https://github.com/openshift/origin/blob/master/pkg/client/client.go#L315 which this now matches.

liggitt commented on hack/common.sh:
```diff
@@ -711,11 +711,11 @@ os::build::ldflags() {
-    ldflags+=(-X "k8s.io/kubernetes/pkg/version.gitCommit=${KUBE_GIT_COMMIT}")
-    ldflags+=(-X "k8s.io/kubernetes/pkg/version.gitVersion=${KUBE_GIT_VERSION}")
+    ldflags+=(-X "k8s.io/kubernetes/pkg/version.gitCommit=${OS_GIT_COMMIT}")
+    ldflags+=(-X "k8s.io/kubernetes/pkg/version.gitVersion=${OS_GIT_VERSION}")
```
what does this do to `oc version` and `openshift version`? those used to spit out origin and k8s versions

deads2k replied:
> what does this do to oc version and openshift version? those used to spit out origin and k8s versions

Messes them up.  Guess I'd rather pull that entirely.  Objections?

liggitt replied:
I vaguely liked having the info available, but I'll let @smarterclayton weigh in

deads2k replied:
> I vaguely liked having the info available, but I'll let @smarterclayton weigh in

Do you have any good advice for making sure that we never accidentally start using after rebases?  Using it in places like command line clients (but only for some resources), `MatchesServerVersion` (but only sometimes), `/version` endpoint?  Mechanics-wise it is painful to keep and of dubious value.  When have you used it?  Was it more useful than knowing the origin level?

smarterclayton replied:
Users and clients want to know what version of Kube we are. We have to
keep it.  We want to get better at describing exactly what version we
include.  We can't really change /version since that is a Kube API

deads2k replied:
> Users and clients want to know what version of Kube we are. We have to
> keep it.  We want to get better at describing exactly what version we
> include.  We can't really change /version since that is a Kube API

Ok, then we have multiple problems:
 1. commit sha points to our fork, not to kube.  This means that a kubectl client will *never* match our server
 1. `/version` returns kube information. This means that an oc client will *never* match our server.

Conclusion: no client ever matches our server.
Conclusion: no one has ever used this because it doesn't work.

Solutions?
 1. respond to `/version` differently based on "User-Agent".
 1. create a second endpoint: `/openshift-versions` that provides enough structure to make a determination
 1. find a way to locate the actual kube tags and commit shas (kind of required)

smarterclayton replied:
I guess I'm ok with /versions reporting our commit.  kubectl will
never match our commit, I mostly care about oc.  I actually don't care
about this feature much - I don't see it *ever* being useful for
anyone in real world.

smarterclayton replied:
I cannot think of a scenario where these are going to match except in RPM installs, and maybe for like 2 days on Online.

liggitt commented on pkg/cmd/server/origin/handlers.go:
```diff
@@ -168,6 +176,98 @@ func namespacingFilter(handler http.Handler, contextMapper kapi.RequestContextMa
+		// user agent is: binary/versionNumber (operating system) openshift/commit
+		// for example: oc/v1.1.3 (linux/amd64) openshift/9b52c41
+		// only match on oc.  This is very restrictive check.  We want to allow by default here.  Clients effectively "opt-in" to
+		// stop stomping things they don't recognize
```
piecemill parsing is super hard to follow... I can't convince you to use a regex here can I?

deads2k replied:
> piecemill parsing is super hard to follow... I can't convince you to use a regex here can I?

`oc/v(\w+) \(.+/.+\) openshift/(\w{7}0`?  can't say that I think its prettier, but if you like it better....

liggitt replied:
I find it easier to see the structure of what is being matched with something like `oc/v(\w+) .* openshift/(\w{7})`, but maybe that's just me

fabianofranz replied:
+1 regex

smarterclayton commented on pkg/cmd/server/origin/handlers.go:
```diff
@@ -168,6 +176,98 @@ func namespacingFilter(handler http.Handler, contextMapper kapi.RequestContextMa
+	if skewedClientPolicy == configapi.AllowAll {
+		return handler
+	}
+	seg := strings.SplitN(binaryInfo.GitVersion, "-", 2)
```
If you need this split it should be done in the build, not here.  This is fragile.

deads2k replied:
This is done upstream to build the user agent string.  Change it there too?  It would be a lot easier to do if the type wasn't unversioned.

So, same logic in both spots, new field duplicating information, or replumb the logic through both sides discrete fields?

smarterclayton replied:
It's fine to have logic in both spots for now, we just need to have an
upstream issue to avoid having to parse.  No one ever spent enough time on
this, really.

On Wed, Mar 2, 2016 at 5:12 PM, David Eads <notifications@github.com> wrote:

> In pkg/cmd/server/origin/handlers.go
> <https://github.com/openshift/origin/pull/7741#discussion_r54809491>:
>
> > @@ -168,6 +176,98 @@ func namespacingFilter(handler http.Handler, contextMapper kapi.RequestContextMa
> >  	})
> >  }
> >
> > +// versionSkewFilter adds a filter that may deny PUT an POST requests since they stomp entire resource objects from skewed
> > +// oc clients, since we know that those clients will strip unknown fields which can lead to unexpected outcomes
> > +func (c *MasterConfig) versionSkewFilter(binaryInfo version.Info, handler http.Handler) http.Handler {
> > +	skewedClientPolicy := c.Options.PolicyConfig.SkewedClientPolicy
> > +	if skewedClientPolicy == configapi.AllowAll {
> > +		return handler
> > +	}
> > +	seg := strings.SplitN(binaryInfo.GitVersion, "-", 2)
>
> This is done upstream to build the user agent string. Change it there too?
> It would be a lot easier to do if the type wasn't unversioned.
>
> So, same logic in both spots, new field duplicating information, or
> replumb the logic through both sides discrete fields?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7741/files#r54809491>.
>

smarterclayton commented on pkg/cmd/server/origin/handlers.go:
```diff
@@ -168,6 +176,98 @@ func namespacingFilter(handler http.Handler, contextMapper kapi.RequestContextMa
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if req.Method != "PUT" && req.Method != "POST" {
```
And PATCH.

deads2k replied:
Not patch.  Patches are non-encoded json that avoids any field stripping logic.  That stripping logic is the only reason to block skewed usage.

smarterclayton replied:
No, skewed usage checks are not only useful for stripping logic, they are
primarily used by admins to force clients to upgrade because the logic the
client is using in the patch is faulty.  The *goal* of skew checks is to
deny clients.  *sometimes* the reason a client is denied is because it
doesn't follow correct logic.

On Wed, Mar 2, 2016 at 5:10 PM, David Eads <notifications@github.com> wrote:

> In pkg/cmd/server/origin/handlers.go
> <https://github.com/openshift/origin/pull/7741#discussion_r54809142>:
>
> > +// versionSkewFilter adds a filter that may deny PUT an POST requests since they stomp entire resource objects from skewed
> > +// oc clients, since we know that those clients will strip unknown fields which can lead to unexpected outcomes
> > +func (c *MasterConfig) versionSkewFilter(binaryInfo version.Info, handler http.Handler) http.Handler {
> > +	skewedClientPolicy := c.Options.PolicyConfig.SkewedClientPolicy
> > +	if skewedClientPolicy == configapi.AllowAll {
> > +		return handler
> > +	}
> > +	seg := strings.SplitN(binaryInfo.GitVersion, "-", 2)
> > +	serverVersionNumber := seg[0][1:]
> > +	serverCommit := binaryInfo.GitCommit
> > +	if len(serverCommit) > 7 {
> > +		serverCommit = serverCommit[:7]
> > +	}
> > +
> > +	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
> > +		if req.Method != "PUT" && req.Method != "POST" {
>
> Not patch. Patches are non-encoded json that avoids any field stripping
> logic. That stripping logic is the only reason to block skewed usage.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7741/files#r54809142>.
>

deads2k replied:
> No, skewed usage checks are not only useful for stripping logic, they are
> primarily used by admins to force clients to upgrade because the logic the
> client is using in the patch is faulty.  The *goal* of skew checks is to
> deny clients.  *sometimes* the reason a client is denied is because it
> doesn't follow correct logic.

This is an outdated diff because you can now choose which verbs you want to apply the check to, so its possible to operate against PATCH.

I still maintain that PATCH is safe and that only really, really, really borked clients will generate patches that valid, but have unintended behavior.

smarterclayton commented on pkg/cmd/server/api/types.go:
```diff
@@ -317,8 +317,24 @@ type PolicyConfig struct {
 	OpenShiftInfrastructureNamespace string
+
+	// SkewedClientPolicy controls how API calls from *voluntarily* identifying clients will be handled.  THIS DOES NOT DEFEND AGAINST MALICIOUS CLIENTS!
+	SkewedClientPolicy SkewedClientPolicy
```
LegacyClientPolicy.   Skewed doesn't have the same meaning to users.

smarterclayton commented on pkg/cmd/server/origin/handlers.go:
```diff
@@ -168,6 +176,98 @@ func namespacingFilter(handler http.Handler, contextMapper kapi.RequestContextMa
+		// only match on oc.  This is very restrictive check.  We want to allow by default here.  Clients effectively "opt-in" to
+		// stop stomping things they don't recognize
+		tokens := strings.SplitN(userAgent, "oc/v", 2)
+		if len(tokens) != 2 {
```
"oc" is not the only client we will need to do this for.  So we should be prepared to expand to multiple client prefixes and types.

deads2k replied:
There's no reason to believe they'd submit the same format of user agent.  The code can be reworked if and when that is required, probably with priority regex matching, it's not particularly dense or layered.

smarterclayton replied:
As mentioned above the point of a skew client check is to give admins a
tool to prevent problems they hit.  Oc isn't the only client - the fact
that we are only worried about it right now is just because it's the first
example, but the actual feature we want is "deny clients based on user
agent".  The selective stripping checks are an optimization for a
particular problem (to be selective), but the more general "deny clients
that report these user agents and give them a clean message" is the feature
we want to offer to admins.

On Wed, Mar 2, 2016 at 5:16 PM, David Eads <notifications@github.com> wrote:

> In pkg/cmd/server/origin/handlers.go
> <https://github.com/openshift/origin/pull/7741#discussion_r54810003>:
>
> > +
> > +		// in pre-1.2 oc clients, the user agent is corrupted when we access kubernetes resources and incorrect versions
> > +		// are passed.  This means that we can't perform any comparison, so we allow this
> > +		// for example: oc/v1.2.0 (linux/amd64) kubernetes/bc4550d
> > +		// that value is passed for a 1.1.3 client when accessing kube resources
> > +		if strings.Contains(userAgent, "kubernetes/") {
> > +			handler.ServeHTTP(w, req)
> > +			return
> > +		}
> > +
> > +		// user agent is: binary/versionNumber (operating system) openshift/commit
> > +		// for example: oc/v1.1.3 (linux/amd64) openshift/9b52c41
> > +		// only match on oc.  This is very restrictive check.  We want to allow by default here.  Clients effectively "opt-in" to
> > +		// stop stomping things they don't recognize
> > +		tokens := strings.SplitN(userAgent, "oc/v", 2)
> > +		if len(tokens) != 2 {
>
> There's no reason to believe they'd submit the same format of user agent.
> The code can be reworked if and when that is required, probably with
> priority regex matching, it's not particularly dense or layered.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7741/files#r54810003>.
>

deads2k commented:
Updated.  `/version` still responds as though it were a strange level of a kube API server.  When accessing kube resources, our client still masquerades as a `kubectl` so we can keep tricking kube API servers (if they care).  When accessing origin resources, our client identifies as openshift.

The server-side filtering now compares provided kube levels against our embedded kube level and provided openshift levels against our openshift level.  I've added a twiddle knob to control the HTTP verbs the skew filter applies to in response to user requests for `GET` prevention.

deads2k commented:
[test]

deads2k commented:
@pweil- ptal.

pweil- commented on pkg/cmd/server/origin/handlers.go:
```diff
@@ -168,6 +178,91 @@ func namespacingFilter(handler http.Handler, contextMapper kapi.RequestContextMa
+	openshiftStyleUserAgent = regexp.MustCompile(`\w+/v([\w\.]+) \(.+/.+\) openshift/\w{7}`)
+)
+
+// versionSkewFilter adds a filter that may deny PUT an POST requests since they stomp entire resource objects from skewed
```
typo: `PUT an POST`

pweil- commented on pkg/cmd/server/origin/handlers_test.go:
```diff
@@ -0,0 +1,245 @@
+	testCases := []versionSkewTestCase{
+		{
+			name:    "missing",
+			methods: verbs,
```
The `missing` cases in the tests need to have a slice with an empty agent to have something to `range` over in order to invoke the tests.

deads2k commented:
comments addressed.  Plan to merge on green test unless someone has something else.

deads2k commented:
[merge]

smarterclayton commented:
Undid merge pending my comments.

deads2k commented:
> The selective stripping checks are an optimization for a
> particular problem (to be selective), but the more general "deny clients
> that report these user agents and give them a clean message" is the feature
> we want to offer to admins.

Do you want to build that full logic now?  I find it difficult to believe that its flexibility we need now.  The code could be extended to take things like user-regexes and display strings for matching purposes, but that's a significant amount of creep for what seems like a really rare case.

smarterclayton commented:
The customers are actually asking for that feature, not this.

On Mon, Mar 7, 2016 at 11:50 AM, David Eads <notifications@github.com>
wrote:

> The selective stripping checks are an optimization for a
> particular problem (to be selective), but the more general "deny clients
> that report these user agents and give them a clean message" is the feature
> we want to offer to admins.
>
> Do you want to build that full logic now? I find it difficult to believe
> that its flexibility we need now. The code could be extended to take things
> like user-regexes and display strings for matching purposes, but that's a
> significant amount of creep for what seems like a really rare case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7741#issuecomment-193367672>.
>

deads2k commented:
One comment seems spurious (PATCH can be added via config), one is a request for a followup issue: https://github.com/openshift/origin/issues/7838.

smarterclayton commented:
Not sure what you mean by spurious, PATCH should default on (unless you've
now changed it so the list is always from config, in which case I'm ok)

On Mon, Mar 7, 2016 at 11:52 AM, David Eads <notifications@github.com>
wrote:

> One comment seems spurious (PATCH can be added via config), one is a
> request for a followup issue: #7838
> <https://github.com/openshift/origin/issues/7838>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7741#issuecomment-193368451>.
>

deads2k commented:
> Not sure what you mean by spurious, PATCH should default on (unless you've
> now changed it so the list is always from config, in which case I'm ok)

The list is always from the config.

smarterclayton commented:
Lgtm, reapply as necessary

deads2k commented:
[merge]
